### PR TITLE
refactor(ec2): MachineImages create appropriate UserData

### PR DESCRIPTION
**Pull request by rix0rrr**
Previously, MachineImages were dumb objects that maybe had UserData
given by the user, and maybe not. If the UserData object was absent,
downstream objects had to inspect the OS type of the MachineImage to
create the right type of UserData object.

That is pretty silly--polymorphism should take care of that unnecessary
`if` there. The MachineImage knows what type of OS it represents, and
should be the one generating the right type of UserData.

Move the responsibility from `Instance` and `AutoScalingGroup` to
`MachineImage`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Commit Message
refactor(ec2): MachineImages create appropriate UserData

Previously, MachineImages were dumb objects that maybe had UserData
given by the user, and maybe not. If the UserData object was absent,
downstream objects had to inspect the OS type of the MachineImage to
create the right type of UserData object.

That is pretty silly--polymorphism should take care of that unnecessary
`if` there. The MachineImage knows what type of OS it represents, and
should be the one generating the right type of UserData.

Move the responsibility from `Instance` and `AutoScalingGroup` to
`MachineImage`.

BREAKING CHANGE: if you implemented a custom subclass of `IMachineImage` it must now always return a `userData` object.

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts [hunk 1]
```diff
@@ -30,10 +30,8 @@ export interface MachineImageConfig {
 
   /**
    * Initial UserData for this image
-   *
-   * @default - Default UserData appropriate for the osType is created
    */
-  readonly userData?: UserData;
+  readonly userData: UserData;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-ec2/lib/machine-image.ts]: This be a breaking change?

File: scripts/builddown [hunk 2]
```diff
@@ -11,4 +11,4 @@ export NODE_OPTIONS="--max-old-space-size=4096 ${NODE_OPTIONS:-}"
 scriptdir=$(cd $(dirname $0) && pwd)
 export PATH=$(cd $scriptdir && npm bin):$PATH
 scope=$(${scriptdir}/current-scope)
-exec lerna run build --scope ${scope} --include-dependents
+exec lerna run ${1:-build} --scope ${scope} --include-dependents
```
Comment (nija-at) [inline scripts/builddown]: Wouldn't you use `foreach` instead?

File: packages/@aws-cdk/aws-ec2/lib/machine-image.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/machine-image.ts]: We don't expect people to instantiate these classes themselves, we expect them to use the existing `MachineImages`. But I'll add a note.

File: scripts/builddown (hunk 2 above)
Comment (rix0rrr) [reply scripts/builddown]: Well that builds everything and this just builds downstream packages.

**Pull request merged**